pkg/middleware/middleware.go

```diff
@@ -346,9 +344,10 @@ func (s *SouinBaseHandler) Upstream(
 			}
 		}
 		s.Configuration.GetLogger().Sugar().Infof("Reused response from concurrent request with the key %s", cachedKey)
-		customWriter.Buf.Reset()
 		customWriter.Buf.Write(sfWriter.body)
-		customWriter.Headers = sfWriter.headers
+		for h, v := range sfWriter.headers {
+			customWriter.Header()[h] = v
+		}
```

Comment:
If using go 1.21, `maps.Copy` is now in the standard library: https://pkg.go.dev/maps#Copy
```suggestion
maps.Copy(customWriter.Header(), sfWriter.headers)
```